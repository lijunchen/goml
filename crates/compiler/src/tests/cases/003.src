fn main() {
    let _ = string_print(unit_to_string(())) in
    let _ = string_print(bool_to_string(true)) in
    let _ = string_print(bool_to_string(false)) in
    let _ = string_print(int_to_string(123)) in
    ()
}
