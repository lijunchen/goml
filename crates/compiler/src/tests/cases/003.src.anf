fn main() -> unit {
  let t4 = unit_to_string(()) in
  let mtmp0 = string_print(t4) in
  let t5 = bool_to_string(true) in
  let mtmp1 = string_print(t5) in
  let t6 = bool_to_string(false) in
  let mtmp2 = string_print(t6) in
  let t7 = int_to_string(123) in
  let mtmp3 = string_print(t7) in
  ()
}