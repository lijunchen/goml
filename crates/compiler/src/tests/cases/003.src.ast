fn main() {
    let _ = print(unit_to_string(())) in
    let _ = print(bool_to_string(true)) in
    let _ = print(bool_to_string(false)) in
    let _ = print(int_to_string(123)) in
    ()
}

