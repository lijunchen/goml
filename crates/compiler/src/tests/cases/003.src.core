fn main () -> Unit {
  let mtmp0 = print_unit(()) in
  let mtmp1 = print_bool(true) in
  let mtmp2 = print_bool(false) in
  let mtmp3 = print_int(123) in
  ()
}