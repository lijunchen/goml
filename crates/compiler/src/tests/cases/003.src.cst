FILE@0..190
  FN@0..190
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..7 "main"
    PARAM_LIST@7..10
      LParen@7..8 "("
      RParen@8..9 ")"
      Whitespace@9..10 " "
    BLOCK@10..190
      LBrace@10..11 "{"
      Whitespace@11..16 "\n    "
      EXPR_LET@16..188
        LetKeyword@16..19 "let"
        Whitespace@19..20 " "
        PATTERN_WILDCARD@20..22
          WildcardKeyword@20..21 "_"
          Whitespace@21..22 " "
        Eq@22..23 "="
        Whitespace@23..24 " "
        EXPR_LET_VALUE@24..50
          EXPR_CALL@24..50
            EXPR_LIDENT@24..29
              Lident@24..29 "print"
            ARG_LIST@29..50
              LParen@29..30 "("
              ARG@30..48
                EXPR_CALL@30..48
                  EXPR_LIDENT@30..44
                    Lident@30..44 "unit_to_string"
                  ARG_LIST@44..48
                    LParen@44..45 "("
                    ARG@45..47
                      EXPR_UNIT@45..47
                        LParen@45..46 "("
                        RParen@46..47 ")"
                    RParen@47..48 ")"
              RParen@48..49 ")"
              Whitespace@49..50 " "
        InKeyword@50..52 "in"
        Whitespace@52..57 "\n    "
        EXPR_LET_BODY@57..188
          EXPR_LET@57..188
            LetKeyword@57..60 "let"
            Whitespace@60..61 " "
            PATTERN_WILDCARD@61..63
              WildcardKeyword@61..62 "_"
              Whitespace@62..63 " "
            Eq@63..64 "="
            Whitespace@64..65 " "
            EXPR_LET_VALUE@65..93
              EXPR_CALL@65..93
                EXPR_LIDENT@65..70
                  Lident@65..70 "print"
                ARG_LIST@70..93
                  LParen@70..71 "("
                  ARG@71..91
                    EXPR_CALL@71..91
                      EXPR_LIDENT@71..85
                        Lident@71..85 "bool_to_string"
                      ARG_LIST@85..91
                        LParen@85..86 "("
                        ARG@86..90
                          EXPR_BOOL@86..90
                            TrueKeyword@86..90 "true"
                        RParen@90..91 ")"
                  RParen@91..92 ")"
                  Whitespace@92..93 " "
            InKeyword@93..95 "in"
            Whitespace@95..100 "\n    "
            EXPR_LET_BODY@100..188
              EXPR_LET@100..188
                LetKeyword@100..103 "let"
                Whitespace@103..104 " "
                PATTERN_WILDCARD@104..106
                  WildcardKeyword@104..105 "_"
                  Whitespace@105..106 " "
                Eq@106..107 "="
                Whitespace@107..108 " "
                EXPR_LET_VALUE@108..137
                  EXPR_CALL@108..137
                    EXPR_LIDENT@108..113
                      Lident@108..113 "print"
                    ARG_LIST@113..137
                      LParen@113..114 "("
                      ARG@114..135
                        EXPR_CALL@114..135
                          EXPR_LIDENT@114..128
                            Lident@114..128 "bool_to_string"
                          ARG_LIST@128..135
                            LParen@128..129 "("
                            ARG@129..134
                              EXPR_BOOL@129..134
                                FalseKeyword@129..134 "false"
                            RParen@134..135 ")"
                      RParen@135..136 ")"
                      Whitespace@136..137 " "
                InKeyword@137..139 "in"
                Whitespace@139..144 "\n    "
                EXPR_LET_BODY@144..188
                  EXPR_LET@144..188
                    LetKeyword@144..147 "let"
                    Whitespace@147..148 " "
                    PATTERN_WILDCARD@148..150
                      WildcardKeyword@148..149 "_"
                      Whitespace@149..150 " "
                    Eq@150..151 "="
                    Whitespace@151..152 " "
                    EXPR_LET_VALUE@152..178
                      EXPR_CALL@152..178
                        EXPR_LIDENT@152..157
                          Lident@152..157 "print"
                        ARG_LIST@157..178
                          LParen@157..158 "("
                          ARG@158..176
                            EXPR_CALL@158..176
                              EXPR_LIDENT@158..171
                                Lident@158..171 "int_to_string"
                              ARG_LIST@171..176
                                LParen@171..172 "("
                                ARG@172..175
                                  EXPR_INT@172..175
                                    Int32@172..175 "123"
                                RParen@175..176 ")"
                          RParen@176..177 ")"
                          Whitespace@177..178 " "
                    InKeyword@178..180 "in"
                    Whitespace@180..185 "\n    "
                    EXPR_LET_BODY@185..188
                      EXPR_UNIT@185..188
                        LParen@185..186 "("
                        RParen@186..187 ")"
                        Whitespace@187..188 "\n"
      RBrace@188..189 "}"
      Whitespace@189..190 "\n"