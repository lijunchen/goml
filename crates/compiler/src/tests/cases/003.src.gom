package main

import (
    "fmt"
)

func unit_to_string(x struct{}) string {
    return "()"
}

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_sub(x int, y int) int {
    return x - y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_print(s string) struct{} {
    fmt.Print(s)
    return struct{}{}
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func missing(s string) struct{} {
    panic("missing: " + s)
    return struct{}{}
}

func main0() struct{} {
    var t4 string = unit_to_string(struct{}{})
    var mtmp0 struct{} = string_print(t4)
    var t5 string = bool_to_string(true)
    var mtmp1 struct{} = string_print(t5)
    var t6 string = bool_to_string(false)
    var mtmp2 struct{} = string_print(t6)
    var t7 string = int_to_string(123)
    var mtmp3 struct{} = string_print(t7)
    return struct{}{}
}

func main() {
    main0()
}
