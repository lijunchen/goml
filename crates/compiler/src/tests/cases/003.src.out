()
====
()truefalse123
====
()truefalse123